bsmconv

0mp opened the pull request.

0mp committed:
Initial commit.

- Implement the reallocation of the buffer within which I will store an
unparsed event.
- Add some notes on how I am going to parse a Linux Audit event.

0mp committed:
Add Makefile.

0mp committed:
Add a simple test script.

0mp committed:
Add .gitignore.

0mp committed:
Fix tests.

0mp committed:
Add records detection.

I can now extract seperate records from the input stream.

0mp committed:
Refactor the code.

0mp committed:
Remove an _ from a variable name.

0mp committed:
Merge pull request #8 from 0mp/bsmconv-refactor

Refactor bsmconv.c

0mp committed:
Update tests.

0mp committed:
Prepare to introduce sbuf(9).

0mp committed:
Use sbuf. (I cannot compile it though.)

Issue:
http://stackoverflow.com/questions/38022644/undefined-reference-to-sbuf-functions-on-freebsd

0mp committed:
Separate the records in eventbuf with EOS.

0mp committed:
Add -pedantic, -ansi, -Werror and -lsbuf flags.

0mp committed:
ANSIfy the code.

0mp committed:
Fix some issues with sbuf.

Some records are still missing from the eventbuf.

0mp committed:
Fix a minor style issue.

0mp committed:
Change the way readbuf is allocated.

0mp committed:
Remove BSMCONV_MSG_FIELD_PREFIX_LEN

0mp committed:
Change iterator variables' names like i to ii.

0mp committed:
Change ++i to i++.

0mp committed:
Improve defines.

0mp committed:
Merge pull request #10 from 0mp/refactor

Refactor

0mp committed:
Change Makefile and add pjdlog files from auditdistd.

0mp committed:
Add a script which adds a path to a dependency for pjdlog.

0mp committed:
Add a pragma to pjdlog.h to ignore not-so-ANSI variadic macros.

Otherwise, `clang -ansi -Werror -Wall ...` refuses to complie
due to `-Wvariadic-macros`.

0mp committed:
Merge pull request #11 from 0mp/compile-pjdlog

Compile pjdlog

0mp committed:
Move makefiles to /usr.bin/bsmconv.

- Remove old Makefile.
- Remove old scripts.
- Remove pragma from pjdlog.h.
- Remove README.

0mp committed:
Merge pull request #12 from 0mp/makefile-to-usrbin

Move makefiles to /usr.bin/bsmconv.

0mp committed:
Remove -l library from Makefile.

0mp committed:
Use pjdlog and update runtest.sh.

0mp committed:
Merge pull request #13 from 0mp/pjdlogging

pjdlogging

0mp committed:
Shorten code lines.

0mp committed:
Improve log messages.

0mp committed:
Fix minor issues and improve tests.

Time to reimplement records and events using TAILQ.

0mp committed:
Remove assert.h from bsmconv.c.

0mp committed:
Update and rename test script.

0mp committed:
Add a successfully initialized TAILQ.

More queues soon.

0mp committed:
Rename the fu script.

0mp committed:
Add extracting of timestamp and id from a record.

0mp committed:
Add tests and modify the fu script.

0mp committed:
Parse timestamp and id. Clean up before parsing fields.

Some tests are still failling.

0mp committed:
Fix the sum function of secs and nsecs.

0mp committed:
Fix test files. (Used ':' instead of '=' in a record's fields).

0mp committed:
Parse record fields. (In progress.)

- Perform s/pjdlog_notice/pjdlog_debug/g.

0mp committed:
Fix pjdlog_debug usage.

0mp committed:
Remove commas from tests. Add a negative comma test.

There should be no commas between records' fields for the time being.

0mp committed:
Improve debug logging. It is possible to control debug level now.

As you normally run `fu t` to run tests you can now run
`fu tv -vvv` to run tests with the debug level of 3.

0mp committed:
Fix fu, the test script.

0mp committed:
Remove unnecessary defines.

0mp committed:
Fix the progress bar in fu, the test script.

0mp committed:
Fix fu, the testing script.

0mp committed:
Fix the sumsecs equation.

0mp committed:
Update and add some tests.

0mp committed:
Solve #21 (the issue with errno != 0).

0mp committed:
Add a test and update the test script.

0mp committed:
Update tests.

0mp committed:
Update tests.

0mp committed:
Parse fields of a record.

I've added a lot of minor changes apart from that.

0mp committed:
Add linau_record_init and solve some TODOs.

0mp committed:
Change mallocs to callocs.

- Incorporate the PJDLOG_VERIFY function.
- Review the correctness of the usage of the PJDLOG_ASSERT function.
  Since asserts can be turned off so use PJDLOG_VERIFY instead of
  PJDLOG_ASSERT if you call any significant functions within the macro.

0mp committed:
Use PJDLOG_VERIFY instead of PJDLOG_ASSERT.

Previously, I checked (bytesread != 1) using an assert which might
result in an uncaught read(2) error.

0mp committed:
Fix minor issues.

0mp committed:
Update string_to_unit32().

Also:
- Change some debug messages.
- Reenable the negative test set.

0mp committed:
Merge pull request #16 from 0mp/tailq

<strike>TAILQ kicks in.</strike>
The idea for the implementation has changed.

0mp committed:
Beta parser (#25)

* Implement a beta version of parsing.

- You can print parsed Linux Audit logs.
- There might be some bugs.
- The code needs to be refactored.

0mp committed:
Add linau_event.c and linau_event.h.

0mp committed:
Parse record's type, id and nsecs.

- I've got to implement the parsing of the record's fields.

0mp committed:
Add gitignore for the makefile directory.

0mp committed:
Change TAILQ to nvlist in the linau structs.

0mp committed:
Use nvlists to manage fields.

- The program compiles and passes the tests.
- Refactoring is badly needed.

0mp committed:
Update tests.

0mp committed:
Rewrite linau_event.

- Use nvlists.
- Standardize the interface.

0mp committed:
Rewrite bsmconv.c.

0mp committed:
Create & design the interface (in progress).

0mp committed:
Refactor and implement the interface (in progress).

0mp committed:
Improve the linau_field interface.

Specifically:
- Fix typedefs.
- Remove some copy-paste code.

0mp committed:
Clean up and fix bugs.

There is a bug in the 452 line of the linau_record when the
linau_record_generate_key tries to get a key.

0mp committed:
Fix a bug in linau_record_generate.

I am about to change the way the key is generated. At the moment the key
is a timestamp contactenated with an id. This is a bad key obviously.

0mp committed:
Remove some old code and add a memo.

0mp committed:
Improve the key generation.

The key of a record is its type.

0mp committed:
Pass tests.

Mainly:
- Update tests.
- Improve interfaces.
- Add more abstract functions to linau_impl.*.

0mp committed:
Implement the event printing.

Generally, the parsing works. There are still some minor issues.

0mp committed:
Implement parser beta version.

Now its time to implement the comma parsing.

Add README with some basic information about the testing script.

Add a test based on read logs.

Fix some minor bugs.

0mp committed:
Add comma support. Update tests.

0mp committed:
Fix the issue with trailing spaces.

0mp committed:
Minor clean-ups.

0mp committed:
Merge pull request #26 from 0mp/library-interface

Library interface

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+
+#include "linau.h"
+
+#include "pjdlog.h"
```
Headers should be organized in groups in the following order:
1. Kernel header files.
2. Standard C Library header files.
3. Other libraries header files.
4. Local header files.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+#include <stdio.h>
+#include <nv.h>
+#include <unistd.h> /* getopt(3) */
```
The comment is unnecessary.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+	linau_record *record;
+	FILE * fp;
+	size_t counter;
+	int debuglevel;
```
Style: wrong variables order. See style(9). There should be no space between an asterisk and a variable name.

0mp replied:
Thanks. I'll try to follow this rule from now on.

Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+	/* Parse command line options. */
+	debuglevel = 0;
+	for (;;) {
+		int ch;
```
Style: variables should be declared at the beginning of the function.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+
+	/* Parse command line options. */
+	debuglevel = 0;
+	for (;;) {
```
I would use the same construction as you can find in getopt(3).

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+
+	pjdlog_debug_set(debuglevel);
+
+	fp = stdin;
```
Let's use file descriptor numbers.

0mp replied:
Like `0` instead of `stidin`?

kwitaszczyk replied:
Yes.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+	counter = 0;
+
+	event = linau_event_create();
+	PJDLOG_VERIFY(event != NULL);
```
PJDLOG_VERIFY should be used to check if a function is successful. I would use PJDLOG_ASSERT here.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+	PJDLOG_VERIFY(event != NULL);
+
+	while ((record = linau_record_fetch(fp)) != NULL) {
+		if (!linau_event_empty(event) &&
```
Why do you need to check if an event is empty? Comparing an origin should be enough.

0mp replied:
You cannot compare origin if the event is empty. Event gets its id and time from the first record in its list.

kwitaszczyk replied:
Maybe it should get an id when it's created?

0mp replied:
@kwitaszczyk It cannot. The id and the time depend on the first record added to an event.

kwitaszczyk replied:
The initial id can be invalid so the comparison would return false it an event is empty.

0mp replied:
Solved. `linau_event_compare_origin` returns 0 if the event is empty.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+		    linau_event_compare_origin(event, record)) {
+			linau_event_print(event);
+			linau_event_destroy(event);
+			event = linau_event_create();
```
You don't check if event is not NULL.

kwitaszczyk replied:
Wouldn't be it better to reset an event instead of allocating it every time there is a new event in a log file?

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,58 @@
+			event = linau_event_create();
+		}
+		linau_event_add_record(event, record, counter);
+		counter++;
```
Why do you need a counter? If you need it for an event specific reason then it should be declared in an event.

0mp replied:
Outdated.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau.h:
```diff
@@ -0,0 +1,8 @@
+
+#include "linau_event.h"
+#include "linau_field.h"
+#include "linau_record.h"
```
It would be better to have all these three structures in one file.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,149 @@
+#include "pjdlog.h"
+
+#include <stdlib.h>
+#include <stddef.h>
```
Style: wrong header files order. See a comment above.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,149 @@
+linau_event_destroy(linau_event *event)
+{
+
+	PJDLOG_ASSERT(event != NULL);
```
Assertions which check input arguments should be separated from the rest of a function with a single blank line.

0mp replied:
I follow this rule already.

Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,149 @@
+	pjdlog_debug(3, " . . + linau_event_add_record");
+	pjdlog_debug(3, " . . . Error (%d)", nvlist_error(event));
+
+	PJDLOG_ASSERT(event != NULL);
```
Assertions which check input arguments should be placed just after variable declarations.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,149 @@
+	    linau_record_get_timestamp(record));
+
+	if (linau_event_empty(event)) {
+		uint64_t timestamp;
```
Style: variables should be declared at the beginning of a function.

0mp replied:
Outdated.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,149 @@
+	uint32_t ide;
+	uint32_t idr;
+	uint64_t tse;
+	uint64_t tsr;
```
These variables should have more intuitive names.

0mp replied:
Solved.

0mp committed:
Refactor library (#32)

0mp committed:
Improve formatting.

0mp committed:
Apply some style(9).

0mp committed:
Clean up.

0mp committed:
Begin to implement linau_event to token conversion.

- Update Makefile.
- Create some basic data structures.
- Add an outline of the linau_event_to_tokenlist function.

0mp committed:
Add a not working version of conversion.

There are still some bugs and the programs doesn't print bytes which
would be understandable to praudit.

0mp committed:
Make a basic conversion work.

0mp committed:
Remove get_size functions.

0mp committed:
Clean up.

0mp committed:
Move the test for tabs to the unhandled tests.

0mp committed:
Expand the libbsm interface.

I need a function like au_close_buffer which takes a timeval structure
and creates header tokens with the time value I provide instead of
using the system time.

0mp committed:
Minor cleanups.

0mp committed:
Comment out au_close_buffer_tm.

0mp committed:
Fix get_timeval.

0mp committed:
Refactor linau_record_to_au. Debugify bsmconv.

0mp committed:
Add LA types defines.

I've got to change them.

0mp committed:
Extend the linau_record structure with lr_text.

0mp committed:
Prepare conversion functions.

- Move defintions to another file.
- Create a function where conversion takes place.
- Delete old functions from linau_record.c.

0mp committed:
Bring back the proper bsmconv.c functionalities.

0mp committed:
Improve the comment of linau_record_compare_origin.

0mp committed:
Update tests. Remove newlines from parsed records.

Additionally, move unhandled tests to .test/.

0mp committed:
Make sense of the Event type parameter in header tokens.

0mp committed:
Add a simple question.

0mp committed:
Update Makefile.

0mp committed:
Modify libbsm/bsm_audit.c and fix bsmconv.c.

Now it is possible to give a custom date to the header token.

0mp committed:
Apply the style(9) rule about the switch/case FALLTHROUGH comment.

0mp committed:
Extend the linau_record interface. Fix a minor bug.

0mp committed:
Add a real test on a record of the USER_CMD type.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,149 @@
+
+	pjdlog_debug(3, " . . . About to generate a key");
+
+	key = malloc((BSMCONV_LINAU_EVENT_KEY_BUFFER + 1) *
```
You don't check if key is NULL.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_field.c:
```diff
@@ -0,0 +1,196 @@
+
+
+static size_t
+find_string_value_end(const char *buf, size_t start, char stringtype)
```
See strstr(3).

0mp replied:
@kwitaszczyk, It is not that simple since this function finds a `stringchar` which doesn't have `\` prepended. That's why this function can find the end of a string like `"name=\"abc def\""` and `strstr` can't.

Right?

kwitaszczyk replied:
I'm not saying you shouldn't use find_string_value_end. Just use strstr internally in this function.

0mp committed:
Merge pull request #37 from 0mp/la2bsmmap

Linux Audit to BSM mapping

0mp committed:
Merge pull request #35 from 0mp/expand-libbsm-interace

Expand the libbsm interface.

0mp committed:
Merge pull request #33 from 0mp/linux-conversion

Linux Audit to BSM conversion (work in progress).

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+#include <unistd.h>
+
+#include <bsm/libbsm.h>
+#include <bsm/audit_record.h>
```
bsm/libbsm.h already includes bsm/audit_record.h.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+#include "pjdlog.h"
+
+#define BSMCONV_BUFFER_SIZE 8192
+
```
This blank space is unnecessary.

0mp replied:
Solved and applied to all files.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+static void	process_event(const struct linau_event *event);
+static void	parse_command_line_options(int argc, char **argv,
+		    int *debuglevelp);
+
```
This blank space is unnecessary.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+#define BSMCONV_BUFFER_SIZE 8192
+
+
+static void	process_events(FILE *fp);
```
Style: there should be one space between a return type and a name of a function.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+static void	process_events(FILE *fp);
+static void	process_event(const struct linau_event *event);
+static void	parse_command_line_options(int argc, char **argv,
+		    int *debuglevelp);
```
Style: wrong indentation, use four spaces for the next line.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+	u_char buf[BSMCONV_BUFFER_SIZE];
+	int aurecordd;
+	struct timeval *tm;
+	unsigned short aueventid;
```
Variables are not sorted.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+
+
+static void
+process_event(const struct linau_event *event)
```
This function should be a part of event's API and should return a buffer.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+}
+
+static void
+process_events(FILE *fp)
```
Use file descriptor numbers.

0mp replied:
@kwitaszczyk The problem is that `fgets` requires `FILE*` not `int`.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+	PJDLOG_ASSERT(fp != NULL);
+
+	event = linau_event_create();
+	PJDLOG_VERIFY(event != NULL);
```
PJDLOG_VERIFY should be used to verify a return value of a function without an assignment. Use PJDLOG_ASSERT here.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/bsmconv.c:
```diff
@@ -0,0 +1,101 @@
+}
+
+static void
+parse_command_line_options(int argc, char **argv, int *debuglevelp)
```
Don't move it to a separate function. It makes code less readable.

kwitaszczyk replied:
Also you call PJDLOG_ABORT here before pjdlog is initialized.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau.h:
```diff
@@ -0,0 +1,120 @@
+
+uint32_t		 linau_event_get_id(const struct linau_event *event);
+uint64_t		 linau_event_get_time(const struct linau_event *event);
+
```
This blank space is unnecessary.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,242 @@
+	struct linau_record *anyrecord;
+
+	PJDLOG_ASSERT(event != NULL);
+	PJDLOG_ASSERT(linau_event_empty(event) == false);
```
!linau_event_empty(event)

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,242 @@
+	PJDLOG_ASSERT(event != NULL);
+	PJDLOG_ASSERT(linau_event_empty(event) == false);
+
+	anyrecord = TAILQ_FIRST(&event->le_records);
```
return (TAILQ_FIRST(&event->le_records));

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,242 @@
+		linau_record_destroy(record1);
+		record1 = record2;
+	}
+	/* XXX Is this really needed? This is what queue(3) says but I don't
```
If you destroy a record but remove from a list then yes, you need it.

0mp replied:
solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,242 @@
+	/* XXX How can I check if the head is initialized? */
+	/* PJDLOG_ASSERT(event->le_records != NULL); */
+
+	/* XXX Should we allow empty events? */
```
Yes.

0mp replied:
OK.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,242 @@
+	int aurecordd;
+
+	PJDLOG_ASSERT(event != NULL);
+	/* XXX How can I check if the head is initialized? */
```
You can assume that if an event exists then the head is initialized.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,242 @@
+
+	/* Tokenise event's records. */
+	TAILQ_FOREACH(record, &event->le_records, lr_next)
+		linau_record_to_au(aurecordd, record);
```
Style: wrong indentation.

kwitaszczyk replied:
This function is a part of record API so the first argument should be a record.

0mp replied:
Style: I believe this indentation is ok because we treat 

```
TAILQ_FOREACH(record, &event->le_records, lr_next)
        linau_record_to_au(aurecordd, record);
```

as

```
for (...)
        linau_record_to_au(...)
```

---

API: Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_event.c:
```diff
@@ -0,0 +1,242 @@
+	TAILQ_FOREACH(record, &event->le_records, lr_next)
+		linau_record_to_au(aurecordd, record);
+
+	*aueventidp = determine_aueventid(event);
```
I wouldn't make it in a separate function.

0mp replied:
solved

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_record.c:
```diff
@@ -0,0 +1,1609 @@
+#define	BSMCONV_LINAU_RECORD_INPUT_BUFFER_SIZE	16
+#define	BSMCONV_LINAU_RECORD_UINT_BUFFER_SIZE	32
+
+#define	LINAU_TYPE_GET_STR			"GET"
```
Move the types to linau_impl.h.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_record.c:
```diff
@@ -0,0 +1,1609 @@
+		PJDLOG_VERIFY(tok != NULL);
+		PJDLOG_VERIFY(au_write(aurecordd, tok) == 0);
+		break;
+	}
```
default is missing.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau.h:
```diff
@@ -0,0 +1,120 @@
+#include <stdio.h>
+
+
+struct linau_event {
```
We should hide these structures' definitions in linau_impl.h and provide an API to access its elements if it's needed.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau_impl.c:
```diff
@@ -0,0 +1,140 @@
+#include <ctype.h>
```
This file should be called linau_common.c.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau.h:
```diff
@@ -0,0 +1,120 @@
+	uint32_t			 lr_id;
+	uint64_t			 lr_time;
+	char				*lr_text;
+	TAILQ_ENTRY(linau_record)	 lr_next;
```
This should be the last element in a structure.

0mp replied:
Solved.

kwitaszczyk commented on contrib/openbsm/bin/bsmconv/linau.h:
```diff
@@ -0,0 +1,120 @@
+
+struct linau_record {
+	char				*lr_type;
+	uint32_t			 lr_id;
```
I would place a record identifier in the beginning of this structure.

0mp replied:
Done.

The previous order originated from the fact that `type` comes before `id` in a Linux Audit record.

kwitaszczyk commented on contrib/openbsm/libbsm/bsm_audit.c:
```diff
@@ -215,12 +215,12 @@ au_write(int d, token_t *tok)
 	struct in6_addr *aptr;
 	struct auditinfo_addr aia;
-	struct timeval tm;
+	struct timeval localtm;
```
I wouldn't change the local variable name but use another name for an argument, let's say eventtm.

0mp replied:
I changed the name to `rectm`.

0mp committed:
Apply suggestions from the code review. (#40)

0mp committed:
Define fields.

0mp committed:
Add a macro which returns the number of arguments.

0mp committed:
Update the macro.

0mp committed:
Extend the linau_record struct with the lr_fields_count field.

0mp committed:
Add a wrapper on pjdlog.h.

0mp committed:
Reorganise the code.

Additionally, I implemented a function which checks if a record has
desired fields.

0mp committed:
Implement a very ugly and incomplete USER_CMD conversion.

0mp committed:
Add a macro to manage token creation.

Additionally:
- Remove mpjdlog. (The problem was with the loglevel changing when you
  called a function inside a function.)
- Add some functions related to conversion.
- The program fails an assertion now. I'll fix it tomorrow.

0mp committed:
Fix a bug which failed the au_write assertion.

0mp committed:
Add a proper text token generation for USER_CMD from the msg field.

0mp committed:
Implement CONVERT_RECORD_TO_AU.

0mp committed:
Improve the file structure.

0mp committed:
Merge pull request #41 from 0mp/la2bsmmap2

LA to BSM mapping

0mp committed:
Add structs.

0mp committed:
Improve the linau_impl interface.

0mp committed:
Implement the basics of the conversion framework.

0mp committed:
Implement the conversion of USER_CMD.

0mp committed:
Update fu and add a js file with LA notes.

0mp committed:
Work in progress.

0mp committed:
Prepare the code for review.

0mp committed:
Merge pull request #43 from 0mp/improve-linau-conv-api

Improve the Linux Audit to BSM conversion API/framework

0mp committed:
Add a new field (see LA docs (ff07afd)).

0mp committed:
Add some comments on tokens and add USER_START conversion.

0mp committed:
Add a CRED_REFR conversion.

0mp committed:
Add a support for a CONFIG_CHANGE conversion.

0mp committed:
Add a support for the USER_END conversion.

0mp committed:
Add support for CRED_DISP.

0mp committed:
Change move functions to set functions.

Solves #42.

0mp committed:
Fix compliation problems.

0mp committed:
Add a macro, fix a bug.

- Fix a bug in functions generating text tokens.
- Add defines to make defining linau_conv_tokens easier.

0mp committed:
Update documentation.

0mp committed:
Add support for the res, result fields.

0mp committed:
Update test log files.

0mp committed:
Update fu.

* Migrate to YAML from JSON.

0mp committed:
Add a return token to DAEMON_START.

0mp committed:
Improve parsing - skip deprecated 'user'.

0mp committed:
Move definitions to a separate file.

0mp committed:
Update fu. Improve ./fu t.

0mp committed:
Move documentation to the docs directory. Use YAML.

0mp committed:
Refactor.

0mp committed:
Add the USER_AUTH test.

0mp committed:
Apply some of Konrad's suggestions (refactoring in progress).

0mp committed:
Simplify the conversion framework.

0mp committed:
Add a TODO file.

0mp committed:
Update docs.

0mp committed:
Make progress bar reset before each test in fu and improve comments.

0mp committed:
Use pjdlog_debug_get and add a todo.

0mp committed:
Improve line breaks (style).

0mp committed:
Perform s/aurecordd/aurd/g.

0mp committed:
Refactor linau_conv_get_type_number.

0mp committed:
Update docs.

0mp committed:
Add some tests for edge cases.

0mp committed:
Remove support for non-space whitespace fields separators.

0mp committed:
Refactor linau_conv.c.

* Rename process_an_id_field to process_id_field.

 The further adjustments have to wait until a major adjustments of
 the static functions' names.

* Improve the name of the linau_conv_write_text_from_field function.

0mp committed:
Update fu.

* Add the help message.
* Refactor the code a little bit.
* Make the progress bar shirnk if you run fu tsv.
* Add colors.
* Extend fu with the ability to generate test cases.

0mp committed:
Add support for every record type in the log files so far.

Additionally:
- Update docs.
- Add a todo.
- Add an additional test for USER_AUTH.

0mp committed:
Refactor the while loop in process_events() in bsmconv.c.

0mp committed:
Add a STYLE file.

0mp committed:
Refactor the code & improve style. Add a todo.

0mp committed:
Improve conversion and style.

Now the the fileds are printed in the desired manner.  Lctokens can only
write desired tokens - they do not care about invalid fields.  It means
that lctokens are not in charge of writing invalid and unprocessed
tokens to the audit record descriptor.

0mp committed:
Update README.md.

0mp committed:
Add some logs generated on cutting-edge CentOS.

0mp committed:
Improve fu (generating fields).

0mp committed:
Support CWD records.

0mp committed:
Update STYLE.

0mp committed:
Improve the path token generation.

0mp committed:
Improve the path token generation.

0mp committed:
Update style guide.

0mp committed:
Add some todos.

0mp committed:
Improve text formatting in fu.

0mp committed:
Update TODO.

0mp committed:
Add tests.

0mp committed:
Add a FreeBSD TODO.

0mp committed:
Improve the process token generation.

0mp committed:
Extend the linau_conv_field interface.

Now there is a union inside the linau_conv_field structure.  It could
be a lcf_validate function like before or a lcf_match function which
returns an nvlist of matching fields.  The point is that it allows us to
detect and handle regex fields.

A regex field is a field the name of which is not predefined as it is
defined with a regex. A regex field is not "a0" while a regex field is
"a4" for example.

0mp committed:
Implement the conversion of EXECVE.

0mp committed:
Try to use the attr token. Fix the path token.

0mp committed:
Improve style (declaration order).

0mp committed:
Add a simple conversion of PROCTITLE.

0mp committed:
Replace string_to_uint32() with an abstraction.

string_to_uint32 wasn't a good function to use on different systems.
I introduced a new function instead.

0mp committed:
Update the TODO file.

0mp committed:
Update the linau_string_queue interface.

0mp committed:
Fix the issue with a fields.

I misunderstood the fact that a0, a1, a2 and a3 are SYSCALL only fields
while a5[2]-like fields are for EXECVE.

0mp committed:
Improve style.

0mp committed:
Update TODO.

0mp committed:
Add the a fields to SYSCALL conversion.

0mp committed:
Add the exit token to SYSCALL.

0mp committed:
Improve conversion, comment on event indentifiers.

 - Add a few tokens to some types.
 - Comment on the indentifiers issue (the
   au_event_type_from_linux_event unimplemented function in
   linux_event.c.

0mp committed:
Add au_to_attr functions to the userland.

0mp committed:
Add the function to convert octal strings to numbers.

0mp committed:
Support the creation of lctoken_attribute.

0mp committed:
Remove an unused function.

0mp committed:
Merge pull request #45 from 0mp/improve-conversion

Improve conversion

0mp committed:
Add a style todo.

0mp committed:
Update the todo list.

0mp committed:
Improve style.

0mp committed:
Remove old docs.

0mp committed:
Add a preamble to the source files.

0mp committed:
Update STYLE.
